services/airflow/dags/helpers/dag_helpers/write_dataframe_to_postgres.py

```diff
@@ -21,60 +21,111 @@
 def connect_to_db_and_insert_pandas_dataframe(
     schema: str, table: str, data: pd.DataFrame
 ):
-    return _connect_to_db_and_insert(_get_tuples_from_pd_dataframe, schema, table, data)
+    return _connect_to_db_and_execute(
+        _build_insert_query(schema, table), _get_tuples_from_pd_dataframe, data
+    )
 
 
 @curry
 def connect_to_db_and_insert_polars_dataframe(
     schema: str, table: str, data: polars.DataFrame
 ):
-    return _connect_to_db_and_insert(
-        _get_tuples_from_polars_dataframe, schema, table, data
+    return _connect_to_db_and_execute(
+        _build_insert_query(schema, table), _get_tuples_from_polars_dataframe, data
     )
 
 
 @curry
-def _connect_to_db_and_insert(
-    tuple_getter, schema: str, table: str, data: pd.DataFrame
+def connect_to_db_and_upsert_pandas_dataframe(
+    schema: str, table: str, constraint: str, data: pd.DataFrame
```

Comment:
```suggestion
    schema: str, table: str, constraint: List[str], data: pd.DataFrame
```

Reply:
should ideally be a list of strings since e.g. ints are not valid sql column names